lib/tests/api/transaction.rs

```diff
@@ -0,0 +1,51 @@
+use surrealdb::sql::{Number, Value};
+use surrealdb::{Connection, Error};
+
+async fn async_query<C>(client: Surreal<C>, sql: &str) -> Result<(), Error>
+where
+	C: Connection,
+{
+	client.query(sql).await?;
+	Ok(())
+}
+
+#[tokio::test]
+async fn verify_transaction_isolation() {
+	// We create two clients
+	let client1 = new_db().await;
+	client1.use_ns(NS).use_db(Ulid::new().to_string()).await.unwrap();
+
+	let client2 = new_db().await;
+	client2.use_ns(NS).use_db(Ulid::new().to_string()).await.unwrap();
+
+	// Create a document with value set to 0
+	client1.query("CREATE foo:bar SET value = 0").await.unwrap();
+
+	// The first client will increment value by 1, but it will be delayed by the slow crypto function.
+	let sql1 = async_query(
+		client1,
+		r#"
+		BEGIN;
+			SELECT * FROM crypto::scrypt::generate("slow");
+			UPDATE foo:bar SET value = value + 1;
+		COMMIT;
+	"#,
+	);
+	// The second client will increment value by 2, without delay.
+	let sql2 = async_query(client2, "UPDATE foo:bar SET value = value + 2");
+
+	// The two queries are run in parallel.
+	let _ = tokio::join!(tokio::spawn(sql1), tokio::spawn(sql2));
+
+	// The final value should be 2 or 1,
+	// Because when both transaction started, the value was 0.
+	// A value of 3 show that the transaction isolation is not respected:
+	// client1 has incremented by 1 the value set by the client2's transaction.
```

Comment:
Cool to see this as a test case! One minor suggestion is that a result of 3 doesn't necessarily mean a snapshot isolation violation under these circumstances, for two reasons:

1. Due to the ambiguities of `async`, client2's transaction could start and finish (incrementing from 0 to 2) before client1's transaction even started (incrementing from 2 to 3). A better test would establish that client1's transaction *definitely* started, then start client2's transaction, then wait for both to finish, and finally test that the result isn't 3. One way to approximate this would be to add a short time delay before spawning client2's transaction.

2. According to my research a moment ago into snapshot isolation, the start/snapshot may be taken _as late as_ the first read (of any key) in the transaction. You could insert a dummy read of any other record before the Scrypt hash (or the same record if you wanted a test case for repeatable read like my second example [here](https://github.com/surrealdb/surrealdb/pull/1529#pullrequestreview-1314588497)). This may make the test more relevant for testing optimized KV stores.

Reply:
Thanks for the feedback @finnbear!
And indeed, it is still a work in progress, as we effectively want to be sure that we detect isolation violations, rather than introducing a flaky test! (Note that the PR is still a draft)

Reply:
@finnbear I finally ended up having a test that follows the script you were suggesting. The result was always 3 (with every engine, including the one having a valid ACID transaction implementation).

I was wondering what I was doing wrong until I decide to check if some of our assumptions could be wrong. And it seems it is the case.

When a transaction starts, it takes a snapshot of the database, ensuring that it is using a consistent database. 
It is true to say that concurrent transactions will then start with the same snapshot of the DB.

But things are a little bit different when concurrent transactions are modifying the same record(s) in a conflicting way. In this case, one transaction (usually with the earlier timestamp) will be allowed to commit its changes, while the transaction(s) with the later timestamp(s) will be rolled back. Eventually, the aborted one(s) is(are) re-applied, but starting with a new snapshot (where `value` is now 1).

Happy to get your thoughts!

Reply:
Cool to hear that the transactions can be automatically retried. I am a bit curious how that works, and whether side effects (such as JavaScript making HTTP requests within the transaction) happen more than once in this case.

As for fixing the test case, I you could test for repeatable read instead of snapshot isolation (see the second example in my comment). To actually test for snapshot isolation, if transactions are being retried, would involve a bit more instrumentation to precisely establish that one transaction started (in the sense that it read something), the other started and the 
 committed a mutation, and finally the first transaction observed that mutation.

Reply:
@finnbear Again thanks for your help here! I have added the test you suggested for checking repeatable reads. To help build more tests in the future, the PR introduces the SQL function Sleep, helpful to write tests with some kind of fair synchronisation. We can also log the execution of the test, providing useful information to check that the sequence has been executed as expected.